code/src/beetle.cpp

```diff
@@ -31,81 +31,163 @@ namespace r2d2::moving_platform {
+                                encode_1_full_turn / wheel_circumference;
+        } else {
+            encoder_rotations = 0;
+        }
+        // The encoder code checks if the pulse goes from low to high. This is
+        // why we start the bool low.
+        bool m0_state = false;
+        uint_fast8_t counter_m0 = 0;
+        uint_fast32_t counter_m0_total = 0;
+        bool m1_state = false;
+        uint_fast8_t counter_m1 = 0;
+        uint_fast32_t counter_m1_total = 0;
+        // The speed of the beetle while moving in a staight line.
+        uint_fast8_t master_power = move_speed;
+        // Initialise slavePower as masterPower - 5 so we don’t get huge error
+        // for the first few iterations
+        uint_fast8_t slave_power = master_power - 5;
+        qik_2s12v10_motorcontroller.set_m1_speed(-slave_power);
+        qik_2s12v10_motorcontroller.set_m0_speed(-master_power);
+        // proportional control setup
+        int_fast8_t error = 0;
+        // proportional_gain 2 worked te best. we only using integers because
+        // calculation with floats on embeded is not that fast.
+        constexpr uint_fast8_t proportional_gain = 2;
+        // sets the tick.
+        uint_fast64_t tick = hwlib::now_us();
+
+        while (counter_m0_total <= encoder_rotations ||
+               counter_m1_total <= encoder_rotations) {
+
+            // encoder for motor 0
+            if (motor_encoder_m0.read() > adc_voltage) {
+                if (m0_state == false) {
+                    counter_m0++;
+                    counter_m0_total++;
+                }
+                m0_state = true;
+            } else {
+                m0_state = false;
+            }
+            // brake the motor if it has reached the target
+            if (counter_m0_total == encoder_rotations) {
+                qik_2s12v10_motorcontroller.brake_m0(0);
+            }
+
+            // encoder for motor 1
+            if (motor_encoder_m1.read() > adc_voltage) {
+                if (m1_state == false) {
+                    counter_m1++;
+                    counter_m1_total++;
+                }
+                m1_state = true;
+            } else {
+                m1_state = false;
+            }
+
+            // brake the motor if it has reached the target
+            // Moter0 has a consistent speed and motor1 has adjustble speed so
+            // it can minimize the error.
+            if (counter_m1_total == encoder_rotations) {
+                qik_2s12v10_motorcontroller.brake_m1(0);
+            } else if (counter_m1_total < encoder_rotations) {
+                // the lowest the motor can power de motor can revieve is 10
+                slave_power = ((slave_power > 10) ? slave_power : 10);
+                qik_2s12v10_motorcontroller.set_m1_speed(-slave_power);
+            }
+
+            // Master slave method
+            // update every interval
+            if (hwlib::now_us() - tick > interval) {
```

Comment:
Tick might not be the best variable name for this it is as far as I know an old time or previous time rather than a tick.

Reply:
previous_measurement_time kudo's to @JasperOosterbroek

Reply:
6db22dfa9e33e1c8a4d681eaa52cad9f30d3d5e2